File: 2. Structural Design Patterns/Composite/Typescript/CompositeTemplate.ts
// Composite Pattern – TypeScript Template

abstract class Component { abstract operation(): void }
class Leaf extends Component { operation(): void { /* ... */ } }
class Composite extends Component {
  private children: Component[] = []
  add(c: Component){ this.children.push(c) }
  remove(c: Component){ this.children = this.children.filter(x=>x!==c) }
  operation(): void { for(const ch of this.children) ch.operation() }
}
(function main(){ const root = new Composite(); root.add(new Leaf()); root.operation() })()

File: 2. Structural Design Patterns/Composite/Kotlin/CompositeTemplate.kt
// Composite Pattern – Kotlin Template

abstract class Component { abstract fun operation() }
class Leaf : Component() { override fun operation() { /* ... */ } }
class Composite : Component() {
    private val children = mutableListOf<Component>()
    fun add(c: Component) = children.add(c)
    fun remove(c: Component) = children.remove(c)
    override fun operation() { children.forEach { it.operation() } }
}
fun main(){ val root = Composite(); root.add(Leaf()); root.operation() }
